Added better filesystem layout testing harness

ranger-ross opened the pull request:
### What does this PR try to resolve?

The goal is to make filesystem layout tests easier to understand and review.
The `build-dir` test's have a [helper fns](https://github.com/rust-lang/cargo/blob/master/tests/testsuite/build_dir.rs#L720) that have imperative checks. While this works, its not easy to see what its validating at a glance.

My thought is to reuse the snapbox infrastructure we already have for the cargo ui tests. 
The prototype in this PR generates a file tree (like the unix tree command) and uses snapbox to compare against a snapshot. This makes the layout tests much easier to read and failures are much more obvious what what went wrong.

### Implementation Details

* The `LayoutTree` struct handles comparison logic (as opposed to using snapbox directly)
  * This avoids the issues initial implementation (ordering and platform specific files, see full details below)
  * If the comparison fails, 2 human readable tree's are generated and snapbox is used to display the diffs. 
    * Note: These diffs are in-memory thus, do not support `SNAPSHOTS=overwrite`. This is arguably good as often the overwrite would be incorrect as the snapshots are platform specific.
* Snapshots would support conditional files based on the target os / platform env
  * Files/directories may be suffixed with `[target_platform=<target-platform>]`
  * Multiple platforms may be specified separated by comma



<details><summary>initial implementation issues</summary>


However there are some problems with the current implementation that limit the usefulness of it.

1. Different platforms have different files which cause some tests to fail
   * Examples
      * lib prefix/suffixes based on platform
      * `dSYM` on macos
   * One idea I have would be to have a cfg suffix after file name in the tree like so
     * `└── foo-[HASH].dSYM [cfg(target_os = "macos")]`
     * This would also require rethinking the "tree lines" (`└`, `├`, `│`) to handle optional files.
2. When dealing with build scripts there are multiple `target/<profile>/build/pkg-[HASH]` directories. The hash changes the order of the directories when generating the tree making snapshots inconsistent.
   * We have redactions to handle replacing the hash with a placeholder `[HASH]`, but this does not help with the order.

</details> 

---

Fun fact: These changes were written and tested at [Rust Forge 2025](https://rustforgeconf.com/) in New Zealand. :new_zealand: :partying_face:

ranger-ross commented on crates/cargo-test-support/src/compare.rs:
```diff
@@ -229,6 +229,11 @@ fn add_regex_redactions(subs: &mut snapbox::Redactions) {
+    // Match multi-part hashes like 06/b451d0d6f88b1d used in directory paths
+    subs.insert("[HASH]", regex!(r"/(?<redacted>[a-f0-9]{2}\/[0-9a-f]{14})"))
+        .unwrap();
+    subs.insert("[HASH]", regex!(r"[a-z0-9]+-(?<redacted>[a-f0-9]{16})"))
```
this is handling the hashes in file/dir names in the fingerprint/build dirs
```
└── debug
    ├── .cargo-lock
    ├── .fingerprint
    │   └── foo-[HASH]
    │       ├── bin-foo
    │       ├── bin-foo.json
    │       ├── dep-bin-foo
    │       └── invoked.timestamp
```

ranger-ross replied:
although I think the first multi-part hash is incorrect now. (I added that when I was still experimenting with a flat file paths)

ranger-ross commented:
I updated the implementation to fix the flaws of the initial design.
See the PR description for the details

The build-dir tests now pass on all of the platforms tested in CI.
I think the changes are now in a state worthy of review.
Let me know what you think

epage commented on crates/cargo-test-support/src/paths.rs:
```diff
@@ -236,6 +241,20 @@ impl CargoPathExt for Path {
+
+    #[track_caller]
+    fn verify_file_layout(&self, expected: impl AsRef<str>) {
+        let actual_layout = LayoutTree::from_path(&self).unwrap();
+
+        let expected_layout = LayoutTree::parse(expected.as_ref());
+
+        if !actual_layout.matches_snapshot(&expected_layout) {
+            let actual_snapshot = actual_layout.to_string();
+            let expected_snapshot = expected_layout.to_string();
+
+            compare::assert_e2e().eq(actual_snapshot, expected_snapshot);
```
> * The `LayoutTree` struct handles comparison logic (as opposed to using snapbox directly)
>       
>   * This avoids the issues initial implementation (ordering and platform specific files, see full details below)
>    * If the comparison fails, 2 human readable tree's are generated and snapbox is used to display the diffs.
>         
>     * Note: These diffs are in-memory thus, do not support `SNAPSHOTS=overwrite`. This is arguably good as often the overwrite would be incorrect as the snapshots are platform specific.
> 
> * Snapshots would support conditional files based on the target os / platform env
>       
>   * Files/directories may be suffixed with `[target_platform=<target-platform>]`
>   * Multiple platforms may be specified separated by comma

Our general aim should be to support snapshotting wherever possible.

Also, tracking of platform-specific details is rough for contributors
- Likely won't discover until the PR is posted
- Likely don't have access to a machine, so must do a slow iteration by pushing and seeing CI's result

Unsure f there is something better but felt this was at least worth specifically discussing

ranger-ross replied:
Yes I totally agree with you.

One idea I have to only check for files in the snapshot and ignore other files. Meaning if a new file was is created a test would not fail. So tests would be able to ignore many of the platform specific files that are not relevant to the test.
But this comes with the trade off of potentially not including a file that is important.

I think I lean toward the current implementation that explicitly matches the file tree structure

epage replied:
Two alternative ideas:
- Have a function that enumerates files in a directory as relative-path-per-line
  - a snapshot assert can then be run against that with the expected results having a `...` inside it and `.unordered()` called on it
  - a parameter could take a glob to ignore
- Have a low level function that accepts a glob to ignore and enumerates files as a tree.  A higher level assert function could assume a specific policy for globs and compare against a snapshot

The problem comes in with
```
[target_platform=windows-msvc]
    │   ├── foo[EXE]
[target_platform=macos,linux,windows-gnu]
    │   ├── foo-[HASH][EXE]   
```

epage replied:
The problem with the existing tests doing `format!("build-dir/debug/deps/foo*{EXE_SUFFIX}")` and if we did `foo[..][EXE]` is that it over matches.  On Unix-like systems, this could match any file and not just the binary.

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -267,17 +476,53 @@ fn examples_should_output_to_build_dir_and_uplift_to_target_dir() {
+    │       └── invoked.timestamp
+    ├── build
+    ├── deps
+    ├── examples
+    │   ├── foo.d                               [target_platform=windows-msvc]
+    │   ├── foo.pdb                             [target_platform=windows-msvc]
+    │   ├── foo[EXE]                            [target_platform=windows-msvc]
+    │   ├── foo-[HASH].d                        [target_platform=macos,linux,windows-gnu]
```
I feel like some of the benefit of `tree` output is lost with the conditionals breaking up the output

ranger-ross replied:
Yeah, it was not very bad until I realized `windows-msvc` does not include hashes in many places that linux, macos, and winddows-gnu does. I felt this was still better than duplicating the tree with  `#[cfg(target_os = "linux")]`.

ranger-ross commented:
I rebased to resolve the conflict with #15910 

I'm waiting to see what we decide to do on this PR before writing tests for the build-dir layout changes.

If we feel this is not a meaningful improvement, I am happy to close this and continue with the existing approach in `testsuite/build-dir.rs`

ranger-ross commented:
In the latest push I made some changes based on the feedback so far:

* Renamed `verify_file_layout()` to `assert_file_layout()`
* Added an `assert_file_layout_with_ignored_paths()` fn that accepts a list of paths that will be ignored
* `assert_file_layout()` adds a predefined list of paths (see: [`default_ignored_paths`](https://github.com/rust-lang/cargo/pull/15874/files#diff-aa8fbecc067fd53b21cfd6c545e92af433478017a604bd9cce844ef0e30c8dc5R339))
  * Currently ignoring platform specific debug symbols on windows and macos that generally get in the way during testing 
* Used wildcard matches (e.g. `foo[..][EXE]`) as mentioned in [this comment](https://github.com/rust-lang/cargo/pull/15874#discussion_r2355759807) 

With these changes the tests are quite a bit cleaner. :) 

I left the `[target_platform=...]` functionality for now as I think it could be potentially useful in the future and is only about 15 lines of code.

epage commented:
Also, while the tree layout works well for inspecting a test, how well does that work for reading the diff and (in rare cases) updating it by hand?

weihanglo commented on src/doc/contrib/src/tests/writing.md:
```diff
@@ -64,6 +64,21 @@ test.
+Tests often to need to verify Cargo created/removed files.
+The `CargoPathExt` trait (implmented by `Path` and `PathBuf`) provides a `assert_file_layout()` to verify a file system tree against a snapshot. The snapshot format is very similar to the unix `tree` command output.
+
+Files vary across operating systems, for example `.pdb` files on Windows. 
```
I wonder instead this platform specific snapshot we should just have different snapshots for different platforms. 

Or maybe we do this kind of tree layout test assertions only on one platform, presumably Linux, which is the most available platforms.
For example, I don't have a Windows physical machine. Some don't have macbook. But everyone can access a Linux container :)

Also, some tier 2 and 3 platform maintainers are running Cargo tests in their CI. While they already patched out some tests, we may want to be nice not to fail their CI more.

weihanglo replied:
Yeah that is why another alternative is testing only on Linux. Anything we really want to test platform differences here?

ranger-ross commented:
> Part of my hope with the rework is we wouldn't need to parse; we generate and then assert. What is left that we still need parsing?

The issue is the file hashes change the order of files/dirs, resulting tests having diff file orders across platforms.
```
    ├── foo-[HASH]
    │        ├── dep-test-integration-test-foo
    │        ├── invoked.timestamp
    │        ├── test-integration-test-foo
    │        └── test-integration-test-foo.json
    └── foo-[HASH]
            ├── bin-foo
            ├── bin-foo.json
            ├── dep-bin-foo
            └── invoked.timestamp
```
Using a tree layout order of lines matters but not the order of the dirs/files.
That is why the `LayoutTree` is used.


> I'd rather stick to what we need than carry around a burden of unused features that exist for speculative purposes, especially if they over encourage the lack of snapshotting.

Sure I can drop this functionality. Its pretty easy to add back if we need it in the future.

> Also, while the tree layout works well for inspecting a test, how well does that work for reading the diff and (in rare cases) updating it by hand?

I found the diffs when the test are wrong are pretty good. 
The 2 pain points I had were, 
* `[EXE]` being removed by mistake in the diff as I was running on linux. Though this is not related to the tree layout.
* At times it was a bit awkward to change a `├` to a `└` and vise versa. Though this was usually just a matter of copy/pasting from another test.

ranger-ross commented:
After some further thought I decided to completely remove `assert_build_dir_layout` and `assert_artifact_dir_layout` in favor of the new `assert_file_layout`. I feel the are largely redundant and file tree snapshots feel like a better approach.

epage commented on crates/cargo-test-support/src/paths.rs:
```diff
@@ -236,6 +249,37 @@ impl CargoPathExt for Path {
+        ignored_paths: &[PathBuf],
+    ) {
+        let assert = assert_e2e();
+        let actual = WalkDir::new(self)
```
Can we enable `sort_by_file_name()`?  Having a more consistent layout can improve the quality of the snapshots and make them easier to browse

epage commented on crates/cargo-test-support/src/paths.rs:
```diff
@@ -290,6 +349,20 @@ where
 }
 
+/// The default paths to ignore when [`CargoPathExt::assert_file_layout`] is called
+fn default_ignored_paths() -> Vec<PathBuf> {
```
Is there a reason this returns `PathBuf`s?  That  makes clear that these are path-like but they aren't ever paths but patterns and this requires the callers to do `unwrap` to undo this.

Maybe the name should be clarified to be about ignored patterns.

ranger-ross replied:
seems reasonable, perhaps `fn default_ignored_path_patterns() -> Vec<String>`

ranger-ross commented on crates/cargo-test-support/src/paths.rs:
```diff
@@ -236,6 +249,37 @@ impl CargoPathExt for Path {
+            })
+            .join("\n");
+
+        assert.eq(format!("{actual}\n"), expected.unordered());
```
sure I like the idea of specialization.
It keeps the tests clean and while removing any specialized build-dir logic from `assert_dir_layout`

epage commented on crates/cargo-test-support/src/paths.rs:
```diff
@@ -236,6 +245,37 @@ impl CargoPathExt for Path {
+
+    #[track_caller]
+    fn assert_build_dir_layout(&self, expected: impl snapbox::IntoData) {
+        self.assert_dir_layout(expected.unordered(), &build_dir_ignored_path_patterns());
```
This should have a comment about why `.unordered()` is needed for build dirs

epage reviewed (approved):
Thanks!
